samcli/commands/destroy/__init__.py

```diff
@@ -0,0 +1,197 @@
+              help="The name of the AWS CloudFormation stack you're deploying to. "
+                   "If you specify an existing stack, the command updates the stack. "
+                   "If you specify a new stack, the command creates it.")
+@click.option('--retain-resources',
+              required=False,
+              multiple=True,
+              default=[],
+              help="For  stacks  in  the DELETE_FAILED state, a list of resource logical"
+                   "IDs that are associated with the resources you want to retain.  During  deletion,  "
+                   "AWS  CloudFormation  deletes  the stack but does not "
+                   "delete the retained resources."
+                   "Retaining resources is useful when you  cannot  delete  a  resource,"
+                   "such as a non-empty S3 bucket, but you want to delete the stack.")
+@click.option('--role-arn',
+              required=False,
+              help="""The Amazon Resource Name (ARN) of an AWS Identity and Access Management (IAM) role that AWS
+              CloudFormation assumes to delete the stack. AWS CloudFormation uses the role's credentials to make
+              calls on your behalf. If you don't specify a value, AWS CloudFormation uses the role  that was
+              previously  associated with the stack. If no role is available, AWS CloudFormation uses a temporary
+              session that is  generated  from your user credentials.""")
+@click.option('--client-request-token',
+              required=False,
+              help="""A unique identifier for this DeleteStack request. Specify this token
+                  if you plan to retry requests so that AWS CloudFormation knows  that
+                  you're  not  attempting  to  delete  a stack with the same name. You
+                  might retry DeleteStack requests to ensure that  AWS  CloudFormation
+                  successfully received them.
+
+                  Learn more at aws cloudformation destroy help
+                  """)
+@click.option('-w', '--wait', required=False, is_flag=True, help="Option to wait for Stack deletion")
+@click.option('--wait-time', required=False, type=click.INT,
+              help="The time to wait for stack to delete in seconds. Used with --wait. The default is 5 minutes")
+@click.option('-y', '--yes', 'ignore_cli_prompt', required=False, is_flag=True,
+              help="Deletes the Stack without cli prompt")
+@common_options
+@pass_context
+@track_command
+def cli(ctx, args, stack_name, retain_resources, role_arn, client_request_token, wait, wait_time=300,
+        ignore_cli_prompt=False):
+    """
+    Destroys the stack
+    """
+    # All logic must be implemented in the `do_cli` method. This helps ease unit tests
+    do_cli(ctx, stack_name, retain_resources, role_arn, client_request_token, wait, wait_time,
+           ignore_cli_prompt)  # pragma: no cover
+
+
+def verify_stack_exists(client, stack_name, required_status=None):
```

Comment:
How about cross stack references? Especially if the stack we are about to delete has exports that are imported in other stacks? how can we be extra careful in those cases?

Reply:
With nested stacks, I can log or output all the stacks that are references or being deleted. Cfn currently deletes the nested stacks and all its references if `aws-cli cloudformation delete-stack` is run. If termination protection is turned on, It seems to only be for the whole nested group. https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/using-cfn-protect-stacks.html